src/wallet/puzzles/did_core.clvm

```diff
@@ -0,0 +1,151 @@
+    )
+  )
+
+  ; takes a lisp tree and returns the hash of it, but allows for escape character 97 so i precomputed hash value can be used
+  (defun sha256tree0 (TREE)
+    (if (l TREE)
+      (if (and (not (l (f TREE))) (= (f TREE) 97))
+          (f (r TREE))
+          (sha256 2 (sha256tree0 (f TREE)) (sha256tree0 (r TREE)))
+      )
+      (sha256 1 TREE)
+    )
+  )
+
+  ; takes a lisp tree and returns the hash of it
+  (defun sha256tree1 (TREE)
+      (if (l TREE)
+          (sha256 2 (sha256tree1 (f TREE)) (sha256tree1 (r TREE)))
+          (sha256 1 TREE)
+      )
+  )
+
+  ;; utility function used by `curry_args`
+  (defun fix_curry_args (items core)
+    (if items
+        (qq (c (q (unquote (f items))) (unquote (fix_curry_args (r items) core))))
+        core
+    )
+  )
+
+  ; (curry_args sum (list 50 60)) => returns a function that is like (sum 50 60 ...)
+  (defun curry_args (func list_of_args) (qq ((c (q (unquote func)) (unquote (fix_curry_args list_of_args (q 1)))))))
+
+  ;; (curry sum 50 60) => returns a function that is like (sum 50 60 ...)
+  (defun curry (func . args) (curry_args func args))
+
+  ;; hash a tree with escape values representing already-hashed subtrees
+  ;; This optimization can be useful if you know the puzzle hash of a sub-expression.
+  ;; You probably actually want to use `curry_and_hash` though.
+  (defun sha256tree_esc_list
+         (TREE LITERALS)
+         (if (l TREE)
+             (sha256 2 (sha256tree_esc_list (f TREE) LITERALS) (sha256tree_esc_list (r TREE) LITERALS))
+             (if (is-in-list TREE LITERALS)
+                 TREE
+                 (sha256 1 TREE)
+             )
+         )
+  )
+
+  ;; hash a tree with escape values representing already-hashed subtrees
+  ;; This optimization can be useful if you know the tree hash of a sub-expression.
+  (defun sha256tree_esc
+     (TREE . LITERAL)
+     (sha256tree_esc_list TREE LITERAL)
+  )
+
+  ;; return the puzzle hash for a cc with the given `genesis-coin-checker-hash` & `inner-puzzle`
+  (defun create_fullpuzhash (mod_hash mod_hash_hash genesis_id inner_puzzle_hash)
+    (sha256tree_esc (curry mod_hash mod_hash_hash genesis_id inner_puzzle_hash)
+                    mod_hash
+                    mod_hash_hash
+                    inner_puzzle_hash)
+  )
+
+  (defmacro create_outputs (puzreveal solution)
+      (qq ((c (unquote puzreveal) (unquote solution))))
+  )
+
+  ; assembles information from the solution to create our own full ID including asserting our parent is a coloured coin
+  (defun-inline create_my_ID (mod_hash mod_hash_hash genesis_id innerpuzhash (parent_parent parent_innerpuz parent_amount) my_amount)
+    (sha256 (sha256 parent_parent (create_fullpuzhash mod_hash mod_hash_hash genesis_id parent_innerpuz) parent_amount) (create_fullpuzhash mod_hash mod_hash_hash genesis_id my_innerpuzhash) my_amount)
+  )
+
+  (defun check_my_amount_and_ID (mod_hash mod_hash_hash genesis_id my_innerpuzhash parent_info my_value)
```

Comment:
Consider using `defun-inline` to see if the code gets smaller. This function is only used once, and all the parameters are "pass-through".